### PR TITLE
Improve X-Forwarded headers

### PR DESCRIPTION
**Pull request by lucaslorentz**
@rdoorn 

I'm creating this PR because I need a way to forward client certificates through mercury.

Changes:
- Added CLIENT_CERT variable with certificate as base64 encoded PEM.
That can be used to create **x-forwarded-client-cert** header.
That's the same approach taken by Azure and Traefik:
https://docs.microsoft.com/bs-cyrl-ba/azure/app-service/app-service-web-configure-tls-mutual-auth?toc=%2Fazure%2Fapp-service-mobile%2Ftoc.json
Traefik
- Added LB_PORT variable with listener's port.
That can be used to create **x-forwarded-port** header.
- Added REQ_PROTO variable with values http or https.
That can be used to create **x-forwarded-proto** header.
- Changed behavior of REQ_IP to return first part of host. It was returning the port.
- Added tests for variable replacement in ACLs

I'm not a golang expert. Let me know what do you think.

### CODE REVIEW

**Comment by lucaslorentz** [comment]
Test configuration:
```
[loadbalancer.pools.LOCAL_IP.listener]
mode = "https"
ip = "172.17.0.2"
port = 443
maxconnections = 9999

[loadbalancer.pools.LOCAL_IP.listener.tls]
certificatekey = "../../test/ssl/self_signed_certificate.key"
certificatefile = "../../test/ssl/self_signed_certificate.crt"

[loadbalancer.pools.LOCAL_IP.backends.postmanecho]
hostnames = ["default"]
connectmode = "http"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.inboundacls]]
action = "add"
header_key = "client-ip"
header_value = "###CLIENT_IP###"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.inboundacls]]
action = "add"
header_key = "req-proto"
header_value = "###REQ_PROTO###"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.inboundacls]]
action = "add"
header_key = "req-host"
header_value = "###REQ_HOST###"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.inboundacls]]
action = "add"
header_key = "lb-port"
header_value = "###LB_PORT###"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.inboundacls]]
action = "add"
header_key = "client-cert"
header_value = "###CLIENT_CERT###"

[[loadbalancer.pools.LOCAL_IP.backends.postmanecho.nodes]]
ip = "34.198.166.230"
port = 80
```

Test 1
```
curl https://172.17.0.2:443/headers --insecure
```
Response 1
```
{"headers":{"x-forwarded-proto":"https","host":"172.17.0.2","accept":"*/*","accept-encoding":"gzip","client-cert":"","client-ip":"172.17.0.2","lb-port":"443","req-host":"172.17.0.2","req-proto":"https","user-agent":"curl/7.58.0","x-forwarded-port":"80"}}
```

Test 2
```
curl https://172.17.0.2:443/headers --insecure --key ./test/ssl/self_signed_certificate.key --cert ./test/ssl/self_signed_certificate.crt --cert-type pem
```

Response 2
```
{"headers":{"x-forwarded-proto":"https","host":"172.17.0.2","accept":"*/*","accept-encoding":"gzip","client-cert":"MIIFwTCCA6mgAwIBAgIJAN1kkxY43QumMA0GCSqGSIb3DQEBCwUAMHcxCzAJBgNVBAYTAk5MMRYwFAYDVQQIDA1Ob29yZC1Ib2xsYW5kMRIwEAYDVQQHDAlBbXN0ZXJkYW0xGDAWBgNVBAoMD1NjaHViZXJnIFBoaWxpczEMMAoGA1UECwwDc2JwMRQwEgYDVQQDDAtleGFtcGxlLmNvbTAeFw0xODAyMDYxMTU3MTJaFw0yODAyMDQxMTU3MTJaMHcxCzAJBgNVBAYTAk5MMRYwFAYDVQQIDA1Ob29yZC1Ib2xsYW5kMRIwEAYDVQQHDAlBbXN0ZXJkYW0xGDAWBgNVBAoMD1NjaHViZXJnIFBoaWxpczEMMAoGA1UECwwDc2JwMRQwEgYDVQQDDAtleGFtcGxlLmNvbTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBAOLcjreSZItdyuKfjjqwYz2ZUTzXM7tEGddF2YjmvZl82OyWhZrX8LyrtHEzmUoesD5ghnlBDGCsv5sUJ58xgbD7bf82LuGdzlP267qf0W6MRmnBs23ZgHetJOtZh+ONaZ7gohnx77JCUR3aSpdgKfBfSBa3T65KnsgpcdYx97r7bfPhlN7DUwkWvAwUKMiDn/dLW0QjVd1tZJXHePNbejMQV1cYaht8m9qbCqt4LyB0Lae5H97Wbi17K3WNvvXJ8Fw7VbLmfF/d+3EECkiBfYY46K6x8VgiQezDt4s/hHRryx3XIjwLsgCZwywrEHTdEcj8VATYrLRTtyzAzZXqOd5SA2nT6hcrbQxUtPf39/wHANL/Tek1u2zvTniRI1MkMy7wxrV17EpHB+5U5/FlYx6m5sbXOoj4fsfN3SsEMBmRiN3Hbk3k5qEybTcMXM390zDiiqYr3BAXGfDoZ9ELY5r5Rb01rd7Z3ckk+ZoFIBshgazEMlOVVX5LnFu5kMGnBcA85CEhZrVeOqVNf0Eex+1GaOCHH2TNeBW/a6C3rbE+64PWPJ7v4KRolVcr/1A/Ma2pLM6/uguTetDPY9Xv7u3IMpKBB1ugB3hCDwQZ+yR3QhjMrUNwNmUmcfbwISFsVkiep10yFDyUpaQ1+mHhiNoDQ2JHPzitAVjLLKjYKfjpAgMBAAGjUDBOMB0GA1UdDgQWBBR8JrkK0JpbkL4/d290Eyj029FjNzAfBgNVHSMEGDAWgBR8JrkK0JpbkL4/d290Eyj029FjNzAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBCwUAA4ICAQBG0PEp4326ovlgHaCKHFmLLSC+w0WvCoxITg2TO8N770lhIOMpQfwohkoN9YFdYtN99PTQODEjxdnnGIYPAvNSn1d6jykPM1zz/S8PsK9BxwmrOGl1+iHv9+ieNTsdfXWZ9y9m7Z15s9IwE4+VRo43/aTBSW5InoZVUeX60KjIifPRsUdgEZz0KPmiWJtaJRMA9Yxgto3SZgg+boDMDVKil9SE3ZJ2nyWiWk3ycA+oGMAjMe84BAGL65ijJXgs/99ilNgwXa5jX+GyXphiqS87wrs6ORw1A53oYituF5YuQ3jF4P0Vk0+XNWF/S5svCKJbnicIZn2ACuG2IdrhnXPjm6xVzmSXm9Pw1u86qlwjD6cUDJwqOC7TNLBDY881TWZaAbE6hK7VUJuDYmZzUb/xG41bbF4k2qVVNprWH4QB44ifWW7mv9WF6zszQjevLbGTgn3pt+pYbAhXPlDqqnGU5A+GYNRPMxgEyZ/U0cGhbWOb1Lm3vmZOzlWs4EjQf9jyGnnNxhsWCNsIxgutWVJSRGkCY4SOeYHhTdbgcvQ23gZ2vVryQKht8FYoEy/KzIDmYPSeGXsoJmpK5s06uJcEnrtgM3a/PID2QcRVtJfgAENujSCyoum2nF3aIxjvRd95vDt9fIvqVRvxD0hsY0+Vo40U5qMscP96v9juFQvTXA==","client-ip":"172.17.0.2","lb-port":"443","req-host":"172.17.0.2","req-proto":"https","user-agent":"curl/7.58.0","x-forwarded-port":"80"}}
```

File: pkg/proxy/proxy_listener.go [hunk 1]
```diff
@@ -109,6 +109,10 @@ func (l *Listener) Start() {
 
 	case "https":
 		proxy := l.NewHTTPProxy()
+
+		// Receive client certificate if provided
+		l.TLSConfig.ClientAuth = tls.RequestClientCert
```
Comment (lucaslorentz) [inline pkg/proxy/proxy_listener.go]: curl handled the certificate request well, even without client certificate. Not sure if any http client would.
Handshake reference:
https://golang.org/src/crypto/tls/handshake_server.go

Comment (rdoorn) [reply pkg/proxy/proxy_listener.go]: By setting this to tls.RequestClientCert, you change the default for all https requests regardless of LB pool, to request a client certificate.
the result is that for all requests, the browser must specificaly send a no certificate message with all requests, creating additional overhead.
this should be a configurable option passed in the config up on loading as part of the tls config provided via the listener

File: pkg/proxy/proxy_http_test.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+var cert = mustParsePEMCertificate(certPEM)
+
+func TestProcessACLVariables(t *testing.T) {
+	//logging.Configure("stdout", "warning")
```
Comment (rdoorn) [inline pkg/proxy/proxy_http_test.go]: logging needs to be configured, or the test will break with a nil pointer dereference
reason being is that the processACLVariables logs errors (such as Unknown variable), and doesn't know how to process the logger yet. - alternatively we prevent processACLVariables from logging errors, and return errors instead, but that requires a larger code change (though it would be a better idea for the future ;-)

File: pkg/proxy/proxy_listener.go (hunk 1 above)
Comment (lucaslorentz) [reply pkg/proxy/proxy_listener.go]: Good point. I created a setting for it.

File: pkg/proxy/proxy_http_test.go (hunk 2 above)
Comment (lucaslorentz) [reply pkg/proxy/proxy_http_test.go]: Since the method was already using logging, I took the short path for now: configuring the logging

**Comment by lucaslorentz** [comment]
Hey @rdoorn 

Your understood my use case is correctly.

Regarding doing TCP load balancing: AFAIK TLS termination of all our components is currently done in mercury. Handling only one component TLS termination outside mercury would require additional setup of certificates in different places. But I will double check with the team what is the preferred option.

Will do the fixes requested

**Comment by lucaslorentz** [comment]
> Thanks for the fixes, one more comment that needs fixing in relation to config changes+reload for your new configuration parameter.

Cool. Didn't know it supported reloading. Fixed this item as well.

Log output after reload with changes to clientauth:
```
DEBU[0073] listener changed - mode:false ip:false port:false, maxcon:false readtimeout:false writetimeout:false ocsp:false cert:false cypher:false curve:false clientauth:true  func=proxy pool=LOCAL_IP tag=core/proxy/init
```

**Comment by armandomagalhesJEAN** [comment]
SOOOO HOW IS IT TO PUSH YOUR SHIT ON OTHERS? ​​​​​​​
HELLO!! LOOKING FOR A GREAT BLOWJOB?? CALL CAROLINE +33661818188 OR VERO +33667301419!! OR CONTACT THEIR SLUTTY FRIEND @IRINAKS1 ;) THEY KNOW HOW TO SUCK WELL!! JEAN BAUDIN KNOWS THAT FOR SURE! HE ENJOYED THE CAROLINE THE MOST !! DON’T BELIVE? CALL HIM AND ASK +31657697635 or +31655935354 XOXO
